rfcs/src/rfcs/0001-mir-linker.md

```diff
@@ -0,0 +1,244 @@
+## Detailed Design
+
+In a nutshell, we will no longer generate a `goto-program` model for every crate we compile.
+Instead, we will generate the `MIR` for every crate, and we will generate only one `goto-program` model.
+This model will only include items reachable from the target crate's harnesses.
+
+The current system flow for a crate verification is the following (Kani here represents either `kani | cargo-kani`
+executable):
+
+1. Kani compiles the user crate as well as all its dependencies.
+   For every crate compiled, `kani-compiler` will generate a `goto-program` model.
+   This model includes everything reachable from the crate's public functions.
+2. After that, Kani links all models together by invoking `goto-cc`.
+   This step will also link against Kani's `C` library.
+3. For every harness, Kani prunes the linked model to only include items reachable from the given harness.
+4. Finally, Kani instruments and verify each harness model via `goto-program` and `cbmc` calls.
+
+After this RFC, the system flow would be slightly different:
+
+1. Kani compiles the user crate dependencies up to the `MIR` translation.
+   I.e., for every crate compiled, `kani-compiler` will generate an artifact that includes the `MIR` representation
+  of all items in the crate.
+2. Kani will generate the `goto-program` only while compiling the target user crate.
+  It will generate one `goto-program` model that includes all items reachable from any harness in the target crate.
+3. `goto-cc` will still be invoked to link the generated model against Kani's `C` library.
+4. Steps #3 and #4 above will be performed without any change.
+
+This feature will require three main changes to Kani which are detailed in the sub-sections below.
+
+### Kani's Sysroot
+
+Kani currently uses `rustup` sysroot to gather information from the standard library constructs.
+The artifacts from this `sysroot` include the `MIR` for generic items as well as for items that may be included in
+a crate compilation (e.g.: functions marked with `#[inline]` annotation).
+The artifacts do not include the `MIR` for items that have already been compiled to the `std` shared library.
+This leaves a gap that cannot be filled by the `kani-compiler`;
+thus, we are unable to translate these items into `goto-program`.
+
+In order to fulfill this gap, we must compile the standard library from scratch.
+This RFC proposes a similar method to what [`MIRI`](https://github.com/rust-lang/miri) implements.
+We will generate our own sysroot using the `-Z always-encode-mir` compilation flag.
+This sysroot will be pre-compiled and included in our release bundle.
+
+We will compile `kani`'s libraries (`kani` and `std`) also with `-Z always-encode-mir`
+and with the new sysroot.
+
+
+### Cross-Crate Reachability Analysis
+
+`kani-compiler` will include a new `reachability` module to traverse over the local and external `MIR` items.
+This module will `monomorphize` all generic code as it's performing the traversal.
```

Comment:
rustc already has logic for this in https://github.com/rust-lang/rust/blob/master/compiler/rustc_monomorphize/src/collector.rs

Maybe it could be generalized to allow you to plug into it and make it work cross crate (rustc only needs cross crate for generic functions and will assume nongeneric functions got codegened upstream)

We have for now given up on mir-only rlibs, but if we want to pick up that work, we'd need exactly what you are describing anyway, so it doesn't seem unreasonable to change the collector to support it optionally.

Reply:
That would be great! Most of the code I've written so far was based on the collector logic.